Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,257 @@
+
+-- We can use implication hypothesis to get new hypotheses
+
+example (hfa : FnUb f a) (hgb : FnUb g b) : FnUb (fun x ↦ f x + g x) (a + b) := by
+  intro x
+  dsimp
+  -- This step makes Lean unpack the definition of the function.
+  -- It is not needed, but it makes things clearer
+  apply add_le_add
+  · exact hfa x
+  -- `hfa x` is now a proof of `f x ≤ a`
+  · exact hgb x
+
+/-
+Since Lean sees an implication as a function from proofs of a Proposition to
+proofs of another Proposition, we can write the proof term explicitly, using `fun`.
+This is more concise, but probably less "human-like." Where to draw the line between
+conciseness and human readability is up to you!
+-/
+
+theorem fnUb_add {f g : ℝ → ℝ} {a b : ℝ} (hfa : FnUb f a) (hgb : FnUb g b) :
+    FnUb (fun x ↦ f x + g x) (a + b) :=
+  fun x ↦ add_le_add (hfa x) (hgb x) -- Note the brackets (terms would be grouped on the left by default)
+
+
+/-!
+# Exercises
+-/
+
+-- Let us define the property a lower bounded function
+def FnLb (f : ℝ → ℝ) (a : ℝ) : Prop :=
+  ∀ x, a ≤ f x
+
+
+-- 1.  Prove this using `intro`
+
+example (hfa : FnLb f a) (hgb : FnLb g b) : FnLb (fun x ↦ f x + g x) (a + b) :=
+  sorry
+
+-- 2. Now try proving it without using any tactic, writing a proof term
+
+example (hfa : FnLb f a) (hgb : FnLb g b) : FnLb (fun x ↦ f x + g x) (a + b) :=
+  sorry
+
+-- 3. Here you may need the theorem `mul_nonneg`
+
+#check mul_nonneg
+
+example (nnf : FnLb f 0) (nng : FnLb g 0) : FnLb (fun x ↦ f x * g x) 0 :=
+  sorry
+
+
+
+
+
+/-!
+  # Existential quantifier
+-/
+
+-- To write `∃` type `\exists`.
+
+/-!
+  To prove an exists statement, we need to provide 2 things. An object of
+  the correct type and a proof that it satisfies the desired property.
+  The first step is done with `use`
+-/
+
+ example : ∃ x : ℝ, 2 * x = 5 := by
+  use 5/2 -- Notice how the goal has changed
+  norm_num -- This is a useful tactic that does arithmetic computations.
+
+-- We can construct a proof term explicitly using the "anonymous constructor" notation.
+-- For this we give the element with the desired property, followed by a proof of the property,
+-- inside angle brackets `⟨⟩` (you can type these using `\<>`).
+
+example : ∃ x : ℝ, 2 * x = 5 :=
+  ⟨5/2 , by norm_num⟩ -- Lean matches the given data with the goal
+
+-- To use an exists statement, we need the command `rcases` and the anonymus constructor
+
+def FnHasUb (f : ℝ → ℝ) :=
+  ∃ a, FnUb f a -- `f` has an upper bound
+
+example (ubf : FnHasUb f) (ubg : FnHasUb g) : FnHasUb fun x ↦ f x + g x := by
+  rcases ubf with ⟨a, ubfa⟩
+  -- Adds new variable `a` (the upper bound) with property `ubfa` (`a` is the upper bound)
+  rcases ubg with ⟨b, ubgb⟩
+  use a + b
+  apply fnUb_add ubfa ubgb
+
+  -- `intro` and `rcases` can be combined into one command: `rintro`
+
+example : FnHasUb f → FnHasUb g → FnHasUb fun x ↦ f x + g x := by
+  rintro ⟨a, ubfa⟩ ⟨b, ubgb⟩
+  exact ⟨a + b, fnUb_add ubfa ubgb⟩
+
+
+/-!
+# Exercises
+-/
+
```

Comment:
```suggestion

-- We open the `Function` namespace to be able to use theorems and definitions about functions with shorter names, e.g. `Surjective f` instead of `Function.Surjective f`.
```